---
title: "Running SCENIC"
subtitle: SCENIC (Single Cell rEgulatory Network Inference and Clustering)
abstract: "First steps for running SCENIC: 
  Filtering the expression matrix, 
  running GENIE3, and 
  (optional) running the next steps in SCENIC using runSCENIC(), the wrapper function."
output:
  BiocStyle::html_document:
    number_sections: no
    toc: yes
  html_notebook:
    toc: yes
  pdf_document:
    toc: yes
package: '`r pkg_ver("SCENIC")`'
bibliography: references.bib
csl: cell.csl
vignette: >
  %\VignetteIndexEntry{Running SCENIC: Step 1 (GENIE3) and 'runSCENIC' (automated steps 2-4)} 
  %\VignetteEngine{knitr::rmarkdown} 
  %\VignetteEncoding{UTF-8}
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE}
# Suppress loading messages when building the HTML
suppressPackageStartupMessages({
    library(Biobase)
    library(data.table)
    library(reshape2)
})

# Do not convert strings to factors (IMPORTANT! Specially if reading-in GENIE3 text output)
options(stringsAsFactors=FALSE)

# To build a personalized report, update this working directory:
knitr::opts_knit$set(root.dir = 'SCENIC_MouseBrain')
```

> This tutorial assumes you have read the introduction to SCENIC and run the setup steps: `vignette(package="SCENIC")`

# Input

## Input format
The input to SCENIC is the single-cell RNA-seq expression matrix, in which each column corresponds to a sample (cell) and each row corresponds to a gene. The gene-symbol should be used as gene ID and stored as `rownames` (for compatibility with RcisTarget annotation databases). Note that in the current version, the motif databases required for RcisTarget are only available for human and mouse. To apply SCENIC to other species, it would require adjustments on the second step.

The preferred expression values are gene-summarized counts (which might or might not use unique molecular identifiers, UMI). Other measurements, such as counts or transcripts per million (TPM) and FPKM/RPKM are also accepted as input. However, note that the first network-inference step is based on co-expression, and some authors recommend avoiding within sample normalization (i.e. TPM) for this task because they may induce artificial co-variation *(@crow2016])*. Although more work is still needed in order to determine the best normalization approach *(@vallejos2017)*, we have applied SCENIC to datasets using raw UMI counts, normalized UMI counts, and TPM and both units provided reliable results (see @aibar2017). The remaining steps of the workflow should not be affected by the expression units/normalization, as the expression is not taken into account for the motif analysis, and AUCell is within-cell ranking-based (wich works as an implicit normalization). 

## Directories
During this workflow we will save many intermediate files. To keep them tidy, we recommend to set the working directory to a new folder. 

Following the workflow *as is*, will save most intermediate files and plots into the `int` folder, with a numbered prefix to keep them in order. The main output (i.e. plots and reports) will be saved into the root folder. If you plan to use the "Wrapper" function to run the steps after GENIE3, you should keep this folder structure and the file names. Otherwise, feel free to modify them to your convenience. 

```{r setwd, results='hide', warning=FALSE, eval=FALSE}
dir.create("SCENIC_MouseBrain")
setwd("SCENIC_MouseBrain") # Or in the first chunk if running a notebook

dir.create("int")
dir.create("output")

# It is also important NOT to convert strings to factors (e.g. when reading from text files)
options(stringsAsFactors=FALSE)
```

# Step 1. Inferring potential TF targets based on co-expression

The first step on SCENIC workflow is to infer potential transcription factor targets based on the expression data. To do this, we use GENIE3. The input to GENIE3 is the expression matrix and a list of transcription factors as *potential regulators*.

In principle, many of the existing methods to infer co-expression networks could be used for this step, as long as its output is a list of potential targets for each TF (and it can be applied to scRNA-seq...).
We selected GENIE3 *(@huynh2010)* because it allows to identify non-linear relationships, even if they are only present in a subset of samples, and it was the best performer in the Network Inference DREAM5 challenge *(@marbach2012)*. 

## Note for big datasets
Running GENIE3 is one of the most time- and computationally-consuming steps of SCENIC. The running time increases with the number of samples (cells). To run SCENIC on large-scale single-cell RNA-seq datasets (i.e. many thousands or millions of cells, depending on the computer infrastructure), we propose two complementary alternatives:

**A.** Using **[GRNboost](https://github.com/aertslab/GRNBoost)**, a scalable framework for gene regulatory network inference using Apache Spark and XGBoost (see @aibar2017)

**B.** **Sub-sampling** the initial dataset (e.g. selecting random or high-quality cells) to infer the co-expression network (either GENIE3 or GRNboost). 
Although the network is *trained* only on subset of cells, if the cells are representative of the dataset, the resulting TF co-expression modules can be used to continue the workflow and infer the gene-regulatory network. In the third step of the workflow (AUCell), the activity of network can be evaluated in all the cells in the dataset. Examples of these approach are presented in see @aibar2017 (i.e. subsampling this mouse brain dataset, and the analysis of 49k cells from mouse retina).


## Load and filter the data
### Load expression matrix
```{r loadExprMat}
# setwd("SCENIC_MouseBrain")
load("data/esetMouseBrain.RData")
exprMat <- exprs(esetMouseBrain)
dim(exprMat)
```

In later steps (step 3 and 4), you can plot some information about the cells on the heatmaps or t-SNE. To use a consistent colouring across all the steps, you may choose some variables (from the phenodata) and colors for each label.

> This information is not required until `runSCENIC()` or steps 3 and 4.

```{r cellInfo, fig.height=4, fig.width=4}
cellInfo <- pData(esetMouseBrain)[colnames(exprMat), "level1class", drop=F]

# Color to assign to the variables (same format as for NMF::aheatmap)
colVars <- list(level1class=setNames(c("forestgreen", "darkorange", "magenta4", "hotpink", "red3", "skyblue", "darkblue"), 
    c("microglia", "endothelial-mural", "astrocytes_ependymal", "oligodendrocytes", "interneurons", "pyramidal CA1", "pyramidal SS")))
save(colVars, file="data/colVars.RData")
plot.new(); legend(0,1, fill=colVars$level1class, legend=names(colVars$level1class))
```

### Select organism

Select the corresponding organism for the databases: mouse ("mm9") or human ("hg19")

```{r chooseOrg}
org <- "mm9"
```

Load corresponding databases:
```{r LoadDbs}
if(org=="hg19")
{
  library(RcisTarget.hg19.motifDatabases.20k)
  
  # Get genes in databases:
  data(hg19_500bpUpstream_motifRanking) # or 10kbp, they should have the same genes
  genesInDatabase <- hg19_500bpUpstream_motifRanking@rankings$rn
  
  # Get TFS in databases:
  data(hg19_direct_motifAnnotation)
  allTFs <- hg19_direct_motifAnnotation$allTFs
}

if(org=="mm9")
{
  library(RcisTarget.mm9.motifDatabases.20k)
  
  # Get genes in databases:
  data(mm9_500bpUpstream_motifRanking) # or 10kbp, they should have the same genes
  genesInDatabase <- mm9_500bpUpstream_motifRanking@rankings$rn
  
  # Get TFS in databases:
  data(mm9_direct_motifAnnotation)
  allTFs <- mm9_direct_motifAnnotation$allTFs
}
```

### Gene filter/selection

To run GENIE3 we recommend to apply soft gene filter to remove genes that are expressed either at very low levels or in too few cells. Here we apply a filtering based on the total number of counts of the gene, and the number of cells in which it is detected.

The first filter, the total number of reads per gene, is meant to remove genes that are most likely noise. 

The second filter, the number of cells in which the gene is detected (e.g. >0 UMI, or >1 log2(TPM)), is to avoid that genes that are only expressed in one, or very few cells, gain a lot of weight if they happen to coincide in a given cell.

To proceed with these filtering, we will first load the matrix, and calculate some basic statistics:

*Number of cells in which each gene is expressed, and number of counts (in the dataset unit) per gene:*
```{r exprStats}
nCellsPerGene <- apply(exprMat, 1, function(x) sum(x>0))
nCountsPerGene <- apply(exprMat, 1, sum)
```

```{r statsSummary}
summary(nCellsPerGene)
summary(nCountsPerGene)
max(exprMat)
sum(exprMat>0) / sum(exprMat==0)
```

**First filter:** We will keep only the genes with at least `r 3*.01*ncol(exprMat)` UMI counts across all samples (e.g. the total number the gene would have, if it was expressed with a value of *3* in 1% of the cells). Adjust this value according to the dataset (it will depend on the dataset units, e.g. UMI, TPMs...).

```{r filter1}
minReads <- 3*.01*ncol(exprMat)
genesLeft_minReads <- names(nCountsPerGene)[which(nCountsPerGene > minReads)]
length(genesLeft_minReads)
```

**Second filter:** We will keep the genes that are detected in at least 1% of the cells. This filtering is meant to remove genes whose reads come from one a few 'noisy' cells. To avoid removing small (but potentially interesting) cell populations, we recommend to set a percentage lower than the smallest population of cells to be detected. 

```{r filter2}
minSamples <- ncol(exprMat)*.01
nCellsPerGene2 <- nCellsPerGene[genesLeft_minReads]
genesLeft_minCells <- names(nCellsPerGene2)[which(nCellsPerGene2 > minSamples)]
length(genesLeft_minCells)
```

**Genes in databases:**  *(Added 31 march 2017)*
In upcoming steps (e.g. motif enrichment analysis), only the genes that are available on RcisTarget databases will be used. To save some running time for GENIE3, we can ignore these genes.
```{r filterGenesInDb}
genesLeft_minCells_inDatabases <- genesLeft_minCells[which(genesLeft_minCells %in% genesInDatabase)]
length(genesLeft_minCells_inDatabases)
```

We can now **filter the expression matrix** to contain only these `r length(genesLeft_minCells_inDatabases)` genes. 
This matrix is now ready to be used as input for GENIE3.
```{r saveFilteredExprMat}
exprMatrix_filtered <- exprMat[genesLeft_minCells_inDatabases, ]
save(exprMatrix_filtered, file="int/1.1_exprMatrix_filtered.RData")
```

```{r checkMissingGenes}
# Check whether any relevant gene / potential gene of interest is missing:
interestingGenes <- c("Neurod1", "Sox10", "Dlx1")
interestingGenes[which(!interestingGenes %in% rownames(exprMatrix_filtered))]
```

To avoid confusions in the following steps: 
```{r rmExprMat}
rm(exprMat)
```

### Potential regulators: List of transcription factors
To provide a list of potential *regulatory genes* to GENIE3, we will use the transcription factors (TFs) that are found in the expression matrix. 
We will identify these using the annotation databases from RcisTarget.

Extract a list of TFs from RcisTarget annotation database:

```{r TFlist}
inputTFs <- allTFs[allTFs%in% rownames(exprMatrix_filtered)]
save(inputTFs, file="int/1.2_inputTFs.RData")

c(allTFs=length(allTFs), inputTFs=length(inputTFs))
```

## Running GENIE3

The input to GENIE3 is the expression matrix, and optionally (but highly recommended), a list of candidate regulators. As candidate regulators, we use the list of transcription factors from RcisTarget databases. 

The filtered expression matrix and the list of TFs (candidate regulators) will be the input to GENIE3. 

GENIE3 default arguments are usually adequate for most datasets. A couple of arguments that might be worth checking are:

- `ncores`: To run GENIE3 in parallel in multiple cores. 

Since GENIE3 is based on a Random Forest approach, each time it is run the results will be slightly different. The higher the number of trees used (`ntrees`), the lower the variability it will be. We recommend to use `set.seed` to reproduce exact results in multiple runs.

For more details, check GENIE3 help with `?GENIE3`.

GENIE3 will typically take several hours (or days) to run. If you are running this workflow on an RStudio session, we recommend that you stop here and run the next code chunk in an independent R console (i.e. with `screen`/`tmux`) or in an server/HPC (if available). The upcoming code chunks will resume the workflow by loading GENIE3 output.

```{r genie3, eval=FALSE}
# setwd("SCENIC_MouseBrain")
load("int/1.1_exprMatrix_filtered.RData")
# Optional: add log (if it is not logged already)
# exprMatrix_filtered <- log2(exprMatrix_filtered+1) 
load("int/1.2_inputTFs.RData")
library(GENIE3)

# Run GENIE3
set.seed(123)
weightMatrix <- GENIE3(exprMatrix_filtered, regulators=inputTFs, ncores=4)
save(weightMatrix, file="int/1.3_GENIE3_weightMatrix.RData")
```

## Correlation

GENIE3 can detect both positive and negative associations. In order to distinguish potential activation from repression, we will split the targets into positive- and negative-correlated targets (i.e. Spearman correlation between the TF and the potential target).

Calculate the correlation (it can be run before/simultaneously to GENIE3):
```{r correlationMat, eval=FALSE}
load("int/1.1_exprMatrix_filtered.RData")
corrMat <- cor(t(exprMatrix_filtered), method="spearman")
save(corrMat, file="int/1.4_corrMat.RData")

# To save storage space, you may save only the rows for TFs:
# corrMat <- corrMat[which(rownames(corrMat) %in% inputTFs),]
```


# Next steps

Once the results from GENIE3/GRNboost (and the correlation) are ready, the remaining steps of SCENIC can be run. You can choose to run each step individually (following the remaining tutorials) or using the "Wrapper function" (`runSCENIC()`, see next section in this tutorial). The wrapper function is equivalent to running each of the individual scripts/tutorias sequentially: 

- Step1.2_CoexpressionModules.Rmd
- Step2_Regulons.Rmd
- Step3.1_NwActivity.Rmd
- Step3.2_BinaryNwActivity.Rmd
- Step4_Clustering.Rmd

The argument `stepsToRun=c("1.2","2", "3.1",  "3.2", "4")` allows to choose the 'scripts' to run. However, no further parameters/options are available. To modify any of them, you can follow the individual script (do not modify the output filenames) and return to the wrapper afterwards. 


If you prefer to use the wrapper function with your own options, feel free to copy and edit the file:
```{r copyRfile, eval=FALSE}
wrapperFile <- paste(file.path(system.file('R', package='SCENIC')), 
                     "SCENIC_wrapper.R", sep="/")
# Copy to edit
file.copy(wrapperFile, ".")
```

## Run SCENIC using the 'Wrapper'
To run the wrapper function, set the working directory (output of GENIE3 will be read from there), load the expression matrix and any information you want to plot from the cells (e.g. color-labels for the heatmap or t-SNE):

```{r loadData2}
# setwd("SCENIC_MouseBrain")
load("data/esetMouseBrain.RData")
exprMat <- exprs(esetMouseBrain)
# Optional: add log for TF expression plot, it does not affect any other calculation
exprMat <- log2(exprMat+1)
dim(exprMat)

load("data/colVars.RData")
cellInfo <- pData(esetMouseBrain)[colnames(exprMat), names(colVars), drop=F]
```

Run the remaining steps using the 'wrapper' function: 
```{r runScenicWrapper, eval=FALSE}
library(SCENIC)
runSCENIC(exprMat=exprMat, org="mm9", cellInfo=cellInfo, colVars=colVars, nCores=4,  stepsToRun=c("1.2", "2", "3.1", "3.2", "4"))
```

# SessionInfo
```{r sessionInfo}
date()
sessionInfo()
```

# References

